lib/modules/dosomething/dosomething_api/resources/magick_resource.inc

```diff
@@ -0,0 +1,144 @@
+            'source' => [
+              'param' => 'campaign',
+            ],
+            'default value' => NULL,
+          ],
+          [
+            'name' => 'campaign_run',
+            'description' => 'The id of the specified campaign run.',
+            'type' => 'string',
+            'optional' => TRUE,
+            'source' => [
+              'param' => 'campaign_run',
+            ],
+            'default value' => NULL,
+          ],
+          [
+            'name' => 'user',
+            'description' => 'The id of the specified user.',
+            'type' => 'string',
+            'optional' => TRUE,
+            'source' => [
+              'param' => 'user',
+            ],
+            'default value' => NULL,
+          ],
+          [
+            'name' => 'key',
+            'description' => 'Custom key to allow automatic campaign signup.',
+            'type' => 'string',
+            'optional' => TRUE,
+            'source' => [
+              'param' => 'key',
+            ],
+            'default value' => NULL,
+          ],
+          [
+            'name' => 'source',
+            'description' => 'Source for the request.',
+            'type' => 'string',
+            'optional' => TRUE,
+            'source' => [
+              'param' => 'source',
+            ],
+            'default value' => NULL,
+          ],
+        ],
+        'access callback' => '_magick_resource_access',
+        'access arguments' => ['index'],
+      ],
+
+    ],
+
+  ];
+
+  return $magick_resource;
+}
+
+/**
+ * Determine whether the method is accessable.
+ *
+ * @return bool
+ */
+function _magick_resource_access($method) {
+  // Temp universal access for now.
+  return TRUE;
+}
+
+/**
+ * Mediate the request and pass along to specfied action if
+ * available.
+ *
+ * @param  string $action
+ * @param  string $campaign
+ * @param  string $user
+ * @return array|ServicesException
+ */
+function _magick_resource_index($action, $campaign, $campaign_run, $user, $key, $source) {
+  if ($action === 'signup') {
+    return magick_signup(compact('campaign', 'campaign_run', 'user', 'key', 'source'));
+  }
+
+  return services_error('This endpoint is specifically for calling custom actions (spells). Please refer to the documentation.', 401);
+}
+
+/**
+ * Complete the automatic campaign signup for a user via
+ * the SignupTransformer class.
+ *
+ * @param  array $parameters
+ * @return array|Exception
+ */
+function magick_signup($parameters) {
+  if (! $parameters['key']) {
+    return services_error('The request is missing the \'key\' token parameter.', 401);
+  }
+
+  if (! $parameters['user']) {
+    return services_error('The request is missing the \'user\' parameter.', 401);
+  }
+
+  if (! $parameters['campaign']) {
+    return services_error('The request is missing the \'campaign\' parameter.', 401);
+  }
+
+  // @TODO: forcing requirement for including the campaign_run, but can likely remove this
```

Comment:
Yeah, I didn't add it in this first attempt because I'd likely need to do that in the transformer, and fully load the user, since `dosomething_helpers_get_current_campaign_run_for_user()` requires a user object and not just the id as well as the fully loaded campaign object. For this first run, I wanted to keep it simple, and eventually extend on it and not require the run id and just let the functions figure it all out.